internal/generate/clusterserviceversion/clusterserviceversion_updaters.go

```diff
@@ -80,36 +81,110 @@ func getCSVInstallStrategy(csv *operatorsv1alpha1.ClusterServiceVersion) operato
 	return csv.Spec.InstallStrategy
 }
 
-// applyRoles updates strategy's permissions with the Roles in the collector.
+// applyRoles applies Roles to strategy's permissions field while respecting ServiceAccount names in RoleBindings.
 func applyRoles(c *collector.Manifests, strategy *operatorsv1alpha1.StrategyDetailsDeployment) {
+	// Collect all role service accounts by their corresponding service accounts via bindings.
+	// This lets us look up all service accounts a role is bound to and create one set of permissions per service account.
+	saNamesToRoleNames := make(map[string]map[string]struct{})
+	for _, binding := range c.RoleBindings {
+		roleRef := binding.RoleRef
+		if roleRef.Kind == "Role" && (roleRef.APIGroup == "" || roleRef.APIGroup == rbacv1.SchemeGroupVersion.Group) {
+			for _, name := range getSubjectServiceAccountNames(binding.Subjects) {
+				if saNamesToRoleNames[name] == nil {
+					saNamesToRoleNames[name] = make(map[string]struct{})
+				}
+				saNamesToRoleNames[name][roleRef.Name] = struct{}{}
+			}
+		}
+	}
+
+	// Apply relevant roles to each service account.
+	unmappedRoles := []rbacv1.Role{}
 	perms := []operatorsv1alpha1.StrategyDeploymentPermissions{}
-	for _, role := range c.Roles {
-		perms = append(perms, operatorsv1alpha1.StrategyDeploymentPermissions{
-			ServiceAccountName: role.GetName(),
-			Rules:              role.Rules,
-		})
+	for saName, roleNames := range saNamesToRoleNames {
+		p := operatorsv1alpha1.StrategyDeploymentPermissions{ServiceAccountName: saName}
+		for _, role := range c.Roles {
+			if _, ok := roleNames[role.GetName()]; ok {
+				p.Rules = append(p.Rules, role.Rules...)
+			} else {
+				unmappedRoles = append(unmappedRoles, role)
+			}
+		}
+		perms = append(perms, p)
 	}
 	strategy.Permissions = perms
+	// Add all unmapped roles to a permission without a service account.
```

Comment:
This isn't the right thing to do (deployment will fail), but should the generator assume in this case that one of the service accounts in a binding can be used here? Is this always the correct thing to do? The default SDK Go project doesn't scaffold a cluster role binding for the [`metrics-reader`](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3/config/rbac/auth_proxy_client_clusterrole.yaml) cluster role.

/cc @joelanford @hasbro17

Reply:
I haven't grokked this PR yet and may not understand the context for your question. So take the following with a grain of salt if you've already thought about this.

The operator doesn't need `metrics-reader`. There are some extra roles that can be bound post-deploy by an admin to enable client interaction with the operator, like this `metrics-reader` one and also the reader and editor roles for each of the CRDs.

I think for any roles that don't have corresponding bindings, maybe we add them to the manifests directory separately, outside the CSV?

Reply:
>we add them to the manifests directory separately, outside the CSV

I'm in favor of doing that.

Reply:
In fact, based on my other comment, I'm kind of thinking that the CSV should only have RBAC rules for the service account used by the deployment.

And all other serviceaccounts, roles, rolebindings, clusterroles, and clusterrolebindings that we slurp in from STDIN are just added to the manifests folder separately.

Reply:
We should probably warn the user if no service account is specified in a deployment's pod template spec then.

Reply:
I think for now mapping (cluster) roles to service accounts using role bindings is ok. When we can patch a deployment to have a pod service account by default, we can change the internal implementation. Thoughts @joelanford?

Reply:
I think its fairly common to not specify `serviceAccountName` in pod specs. And I think that's totally okay. 

If we see that `spec.serviceAccountName == ""`, that's equivalent to `spec.serviceAccountName == "default"` so when we're looking at role bindings, we look for service account "default" when the spec service account is empty.